### ISSUE TITLE
CPU utilization metrics wildly inaccurate

### ISSUE THREAD

**Issue by phemmer** [kubernetes/heapster#759]
I'm new to heapster, influxdb, grafana, so I apologize if this isn't the right place, but this seems like an issue with heapster.

The problem is: I have an app running inside a pod which is consuming about 40% CPU (according to `docker stats` on the container), though it does bounce up and down a bit. However when I look at the pod in grafana, the chart is all over the place, including showing negative values. And the spikes grow larger over time.

![image](https://cloud.githubusercontent.com/assets/1826947/11544044/d1bd8fda-990d-11e5-9cf4-7ba93707f01f.png)

Docker stats:

```
CONTAINER           CPU %               MEM USAGE/LIMIT     MEM %               NET I/O
a990                46.93%              8.126 MB/3.951 GB   0.21%               0 B/0 B
```

&nbsp;

Version stuff:
```
# kubectl version
Client Version: version.Info{Major:"1", Minor:"2+", GitVersion:"v1.2.0-alpha.4", GitCommit:"4a9b0fc7153c733832d70c9f6d0796db177643a6", GitTreeState:"clean"}
Server Version: version.Info{Major:"1", Minor:"1", GitVersion:"v1.1.1", GitCommit:"92635e23dfafb2ddc828c8ac6c03c7a7205a84d8", GitTreeState:"clean"}

# kubectl get rc heapster influxdb-grafana --namespace=kube-system
CONTROLLER         CONTAINER(S)   IMAGE(S)                             SELECTOR                      REPLICAS   AGE
heapster           heapster       kubernetes/heapster:canary           k8s-app=heapster,version=v6   1          1h
influxdb-grafana   influxdb       kubernetes/heapster_influxdb:v0.5    name=influxGrafana            1          1h
                   grafana        kubernetes/heapster_grafana:v2.1.0

# kubectl get pod heapster-8xek5 --namespace=kube-system -o jsonpath='{.status.containerStatuses[].imageID}'
docker://0a433ef5a50163c7b1552d7a478fee995f23f8679d853a8e1e9ea877900ce36e
```

**Comment by vishh** [comment]
This is an issue with older versions of InfluxDB. A more recent version,
InfluxDB v0.9.x, does not have this issue. The next heapster release will
include support for v0.9.x and it should be released this week.

**Comment by antoineco** [comment]
On a stack running `heapster:v0.19.1`, `influxdb:v0.6` (v0.9.6) and `grafana:v2.5.0`, CPU metrics are even simply absent:

<img src="https://cloud.githubusercontent.com/assets/3299086/11962527/c5286596-a8e0-11e5-9204-b0879107ecee.png" height="200">

@vicki-c: I'm using the dashboards @master ([here](https://github.com/kubernetes/heapster/tree/aa5a129d2852ba140a8c68c9aa58c1bfe863d8cd/grafana/dashboards)) and my assumption is that a "Group by time interval" of 10s might be too low for most use cases. If I replace set "Group by time interval" to `> 30s` I seem to get more meaningful results:

<img src="https://cloud.githubusercontent.com/assets/3299086/11963008/405bd34e-a8e4-11e5-8026-fb3bda0202ad.png" height="200">

Not sure if that makes sense for every kind of app though.

**Comment by antoineco** [comment]
This is not yet sufficient, the patterns look right but the CPU usage calculation doesn't.

--

Comparison of CPU metrics from New Relic, Kubedash and Grafana for one single node over 1 hour:

<img src="https://cloud.githubusercontent.com/assets/3299086/11973798/6a607d1e-a95e-11e5-868c-c6d5ac7d33ed.png" width="500px">
<img src="https://cloud.githubusercontent.com/assets/3299086/11973796/6a5e9f76-a95e-11e5-904d-1321c48a4240.png" width="500px">
<img src="https://cloud.githubusercontent.com/assets/3299086/11973797/6a5f9408-a95e-11e5-9198-ddbdd77fa0b5.png" width="500px">

--

This results in a displayed value nearing or above 100% (8000m, in my case), which is most likely wrong:

<img src="https://cloud.githubusercontent.com/assets/3299086/11973988/0f4af6aa-a960-11e5-8079-ed2fcfaf8b7a.png" width="500px">
<img src="https://cloud.githubusercontent.com/assets/3299086/11973989/0f6c38b0-a960-11e5-82fb-480b75f0492f.png" width="500px">

--

Fiddling with the "Group by interval" and changing it from `>30s` to precisely `16s` in this case gives me the following results:

<img src="https://cloud.githubusercontent.com/assets/3299086/11974075/fcebd7ee-a960-11e5-8e94-193644418ef7.png" width="500px">

The higher the interval, the higher the value, which sounds wrong from the user's perspective. Here with 120s:

<img src="https://cloud.githubusercontent.com/assets/3299086/11974112/65d0e736-a961-11e5-8075-2af0f9bd2ebb.png" width="500px">

--

I'll investigate further how we could determine something accurate regardless of the chosen interval. @vishh should I create a new issue or do you prefer to re-open that one?

**Comment by vishh** [comment]
Either works for me. Thanks for investigating!

**Comment by antoineco** [comment]
I found the issue and pushed the changes as part of #824

**Comment by stefanodoni** [comment]
Hi,

I have the same problem,

I see negative CPU usage in Grafana dashboard:

![image](https://cloud.githubusercontent.com/assets/15234197/15246757/918ffd62-190f-11e6-9630-4c5f673b7c4f.png)

Here is kubernetes version:

$ kubectl version
Client Version: version.Info{Major:"1", Minor:"2", GitVersion:"v1.2.2", GitCommit:"528f879e7d3790ea4287687ef0ab3f2a01cc2718", GitTreeState:"clean"}
Server Version: version.Info{Major:"1", Minor:"2", GitVersion:"v1.2.3+coreos.0", GitCommit:"c2d31de51299c6239d8b061e63cec4cb4a42480b", GitTreeState:"clean"}

As regard the monitoring, I have used the YAML bundled with kubernetes 1.2.4 official tar, installed with:

$ kubectl create -f addons/cluster-monitoring/influxdb

I'm having hard time to get a correctly working heapster + grafana + influxdb setup. The official monitoring guide (https://github.com/kubernetes/heapster/blob/master/docs/influxdb.md) yelds a setup where Grafana is not showing anything, due to wrongly named metrics (e.g. memory/limit_gauge_something). I have then used the YAML included as part of the official K8S release, but I'm still getting wrong values.

Am I missing something?

Thank you in advance for any help.
